### PR TITLE
Add `RZ_BUILD_DEBUG` compile time definition for debug logging

### LINKED ISSUE #1612: RZ_LOG_DEBUG should not be compiled in releases mode
**Is your feature request related to a problem? Please describe.**
While debugging/developing you may want to add a lot of logs that can be enabled with `e log.level=X` where X is a number for 0 to 6. However, when we release Rizin we likely do not want those DEBUG prints to ever be printed, saving in some cases a lot of useless printing and calculations.

**Describe the solution you'd like**
When buildtype==release, RZ_LOG_DEBUG and RZ_LOG_SILLY should be defined to nothing, thus they will be compiled out entirely from release builds.

**Describe alternatives you've considered**
None.

**Additional context**
None.

### PR DESCRIPTION
**Pull request by DMaroo**
**Your checklist for this pull request**
- [x] I've read the [guidelines for contributing](https://github.com/rizinorg/rizin/blob/master/DEVELOPERS.md) to this repository
- [x] I made sure to follow the project's [coding style](https://github.com/rizinorg/rizin/blob/master/DEVELOPERS.md#code-style)
- [x] I've added tests that prove my fix is effective or that my feature works (if possible)
- [x] I've updated the documentation and the [rizin book](https://github.com/rizinorg/book) with the relevant information (if needed)

**Detailed description**

`RZ_LOG_DEBUG` and `RZ_LOG_SILLY` are now only compiled when `RZ_BUILD_DEBUG` is set/defined.

**Test plan**

Green CI.

**Closing issues**

Closes #1612.

### CODE REVIEW

**Comment by DMaroo** [comment]
Do I need to add tests for this?

**Review by ret2libc — CHANGES_REQUESTED** [review]
Instead of adding a new variable I'd suggest to use NDEBUG define.

You can do #ifdef NDEBUG and use add `default_options: ['b_ndebug=if-release']` in the `project` part of the main meson.build file.

**Comment by ret2libc** [comment]
> Do I need to add tests for this?

It should not be needed. We already have several CI jobs that use both debug and release buildtypes.

**Comment by DMaroo** [comment]
> Instead of adding a new variable I'd suggest to use NDEBUG define.
> 
> You can do #ifdef NDEBUG and use add `default_options: ['b_ndebug=if-release']` in the `project` part of the main meson.build file.

Since we are using our own logging functions, should we avoid using `NDEBUG`, since defining/un-defining it will affect other things as well, like assert macros? Just asking, I don't know which of the ways is preferred.

**Comment by ret2libc** [comment]
> > Instead of adding a new variable I'd suggest to use NDEBUG define.
> > You can do #ifdef NDEBUG and use add `default_options: ['b_ndebug=if-release']` in the `project` part of the main meson.build file.
> 
> Since we are using our own logging functions, should we avoid using `NDEBUG`, since defining/un-defining it will affect other things as well, like assert macros? Just asking, I don't know which of the ways is preferred.

Yeah maybe you are right... We could also re-use the RZ_CHECKS_LEVEL define that we already have, which is 1 on release builds and 2 otherwise.

```
/*
 * RZ_CHECKS_LEVEL determines the behaviour of the rz_return_* set of functions.
 *
 * 0: completely disable every function and make them like no-operation
 * 1: silently enable checks. Check expressions and do return, but do not log anything
 * 2: enable checks and logging (DEFAULT)
 * 3: transform them into real assertion
 */
```
In that case, we should make RZ_LOG_DEBUG/SILLY no-op if RZ_CHECKS_LEVEL <= 1.

Or we could just use the additional define you did :D Not really so sure anymore...

@kazarmy @XVilka ?

**Comment by DMaroo** [comment]
Also, since we are using `-Werror` in the Linux gcc build, therefore, warnings such as `unused variable` are leading to failure (https://github.com/rizinorg/rizin/pull/1616/checks?check_run_id=3533582589). This is introduced because `RZ_DEBUG_LOG` becomes a no-op, and therefore, some of the variables are used nowhere. I think we can just do some sort of `(void)fmtstr` and same for rest of the arguments to avoid the warning.

**Comment by ret2libc** [comment]
> Also, since we are using `-Werror` in the Linux gcc build, therefore, warnings such as `unused variable` are leading to failure (https://github.com/rizinorg/rizin/pull/1616/checks?check_run_id=3533582589). This is introduced because `RZ_DEBUG_LOG` becomes a no-op, and therefore, some of the variables are used nowhere. I think we can just do some sort of `(void)fmtstr` and same for rest of the arguments to avoid the warning.

Yeah that should be easily fixed. Only few cases, it seems so far.

File: librz/type/parser/c_cpp_parser.c [hunk 1]
```diff
@@ -230,6 +230,7 @@ RZ_API int rz_type_parse_file_stateless(RzTypeParser *parser, const char *path,
 		return -1;
 	}
 	ut64 file_size = rz_file_size(path);
+	(void)file_size;
 	RZ_LOG_DEBUG("File size is %" PFMT64d " bytes, read %zu bytes\n", file_size, read_bytes);
```
Comment (wargio) [inline librz/type/parser/c_cpp_parser.c]:
```suggestion
	RZ_LOG_DEBUG("File size is %" PFMT64d " bytes, read %zu bytes\n", rz_file_size(path), read_bytes);
```

File: librz/type/parser/c_cpp_parser.c [hunk 2]
```diff
@@ -252,6 +253,7 @@ RZ_API int rz_type_parse_file(RzTypeDB *typedb, const char *path, const char *di
...
```
Comment (wargio) [inline librz/type/parser/c_cpp_parser.c]:
```suggestion
	RZ_LOG_DEBUG("File size is %" PFMT64d " bytes, read %zu bytes\n", rz_file_size(path), read_bytes);
```

**Review by wargio — COMMENTED** [review]
How about this way?

File: meson.build [hunk 3]
```diff
@@ -298,6 +298,10 @@ if checks_level == 9999
   endif
 endif
 
+if get_option('buildtype').startswith('debug')
+  add_project_arguments('-DRZ_BUILD_DEBUG', language : 'c')
+endif
```
Comment (ret2libc) [inline meson.build]: Let's put this thing in rz_userconf.h.in and set it similar to the other variables there. That way people using the API can also see if rizin was compiled in debug mode or not.

File: librz/type/parser/c_cpp_parser.c (hunk 1 above)
Comment (DMaroo) [reply librz/type/parser/c_cpp_parser.c]: Makes sense to me. Thanks.

**Review by XVilka — APPROVED** [review]

**Comment by XVilka** [comment]
I think it would be worth to cherry-pick it into 0.3.0 as well, to reduce the size of the binaries (slightly).

**Review by ret2libc — APPROVED** [review]

**Pull request merged**